example/l2fwd/README

```diff
@@ -5,3 +5,7 @@ Source code and Makefiles placed under test/performance/ directory.
 This L2 forwarding application can be used as example reference as well
 as performance test for different odp modes (direct, queue or scheduler
 with parallel, atomic or ordered queues).
+
+Note that this example is tuned for performance. As a result when using
+ordered queues with direct packet I/O output, packet order is not
+guaranteed. To achieve guaranteed order, use queued output (-o option).
```

Comment:
Using queued output is not sufficient in case of l2fwd since the l2fwd may send packets received from the same input queue to different output queues of the same output port. This is because the output queue is selected based on the thread that happens to be handling the packet and not e.g. based on flow hash or the input queue. For the same reason, order is not necessarily preserved with atomic queues either (enqueuing would happen in order with atomic queues, but packets in more than one queue may pass each other before output to the wire).

The only case where l2fwd guarantees that the input order is preserved in output is when there is only one worker thread.

Reply:
Yes, having only one output queue per interface would preserve the order, but that would indeed require some implementation work. Multiple output queues per interface with a flow based or input queue based output queue selection would also preserve the order.